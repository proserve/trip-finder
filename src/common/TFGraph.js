/* eslint-disable no-param-reassign, no-loop-func */

import get from 'lodash/get';
import set from 'lodash/set';

export const weightProps = {
  MONEY: 'MONEY',
  TIME: 'TIME',
};

class TFGraph {
  constructor(deals, departure, arrival, weightProp = weightProps.MONEY,
    filters = { bus: true, train: true, car: true }) {
    this.deals = deals;
    this.departure = departure;
    this.arrival = arrival;
    this.filters = filters;
    // TODO: throw error in case of
    this.weightProp = weightProp; // possible values (distance, time)
    this.createGraph = this.createGraph.bind(this);
    this.graph = this.createGraph();
  }

  getWeight({ cost, discount, duration: { h, m } }) {
    if (this.weightProp === weightProps.TIME) {
      return parseInt(h, 0) * 60 + parseInt(m, 0);
    }
    if (this.weightProp === weightProps.MONEY) {
      return cost - (cost * (discount || 0) / 100);
    }
    return 0;
  }

  createGraph() {
    const graph = {};
    this.deals.forEach(({
      departure, arrival, cost, discount, transport, reference, duration,
    }) => {
      if (this.filters[transport]) {
        const key = `${departure}.${arrival}`;
        let weight = this.getWeight({ cost, discount, duration });
        const currentWeight = get(graph, `${key}.weight`) || Infinity;
        const newTrans = currentWeight > weight ? transport : get(graph, `${key}.transport`);
        const newRef = currentWeight > weight ? reference : get(graph, `${key}.reference`);
        const price = currentWeight > weight ? cost : get(graph, `${key}.price`);
        const cDiscount = currentWeight > weight ? discount : get(graph, `${key}.discount`);
        const cDuration = currentWeight > weight ? duration : get(graph, `${key}.duration`);
        if (currentWeight) {
          weight = Math.min(weight, currentWeight);
        }
        set(graph, key, {
          weight,
          transport: newTrans,
          reference: newRef,
          price,
          discount: cDiscount,
          duration: cDuration,
        });
      }
    });
    return graph;
  }

  static findCheapestCost(costs) {
    return Object.keys(costs).reduce((cheapest, city) => {
      const cheapestCost = get(costs, `${cheapest}.cost`) || Infinity;
      if ((cheapest == null || costs[city].cost < cheapestCost)
          && !costs[city].visited) {
        cheapest = city;
      }
      return cheapest;
    },
    null);
  }

  calculateCheapestTrip() {
    if (!this.departure || !this.arrival || this.departure === this.arrival) {
      return alert('Please enter departure and arrival');
    }
    const costs = {
      [this.departure]: {
        cost: 0,
        visited: true,
        parent: null,
        transport: null,
        reference: null,
      },
      [this.arrival]: {
        cost: Infinity,
        visited: true,
        parent: null,
        transport: null,
        reference: null,
      },
    };
    const startNode = this.graph[this.departure];
    Object.keys(startNode).forEach((city) => {
      const {
        transport, reference, price, discount, duration,
      } = startNode[city];
      costs[city] = {
        cost: startNode[city].weight,
        visited: false,
        parent: this.departure,
        transport,
        reference,
        price,
        discount,
        duration,
      };
    });
    let cheapest = TFGraph.findCheapestCost(costs);
    while (cheapest) {
      const currentNode = this.graph[cheapest];
      Object.keys(currentNode).forEach((city) => {
        this.updateNodeValue(costs, currentNode, city, cheapest);
      });
      costs[cheapest].visited = true;
      cheapest = TFGraph.findCheapestCost(costs);
    }

    return this.calculateResult(costs);
  }

  calculateResult(costs) {
    if (costs[this.arrival].parent == null || costs[this.arrival] === Infinity) {
      return [];
    }
    const result = [{ ...costs[this.arrival], current: this.arrival }];
    let cNode = costs[this.arrival];
    while (cNode.parent) {
      result.push({ ...costs[cNode.parent], current: cNode.parent });
      cNode = costs[cNode.parent];
    }
    result.pop();
    return result.reverse();
  }

  updateNodeValue(costs, currentNode, city, cheapest) {
    const arrivalCost = get(costs, `${this.arrival}.cost`);
    const { weight } = currentNode[city];

    let currentCost = get(costs, `${city}.cost`);
    currentCost = currentCost === undefined ? Infinity : currentCost;

    const newCost = weight + costs[cheapest].cost;

    const cost = Math.min(currentCost, newCost);
    // Since all distances and time are positive no need to continue with path
    // that cost more than the current arrivalCost
    if (cost < arrivalCost) {
      const visited = currentCost < newCost && costs[city].visited;
      const parent = currentCost < newCost ? costs[city].parent : cheapest;
      const {
        transport, reference, duration, price, discount,
      } = currentCost < newCost ? costs[city] : currentNode[city];
      costs[city] = {
        cost, visited, parent, transport, reference, duration, price, discount,
      };
    }
  }
}

export default TFGraph;
