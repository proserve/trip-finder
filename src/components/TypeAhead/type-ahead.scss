@import '../../common/style/variable';

.type-ahead-container {
  position: relative;
  float: left;
  margin-bottom: 5px;
  min-width: 250px;

  input {
    padding: 0 10px;
    border: 1px solid #EBEBEB;
    font-size: 14px;
    height: 48px;
    border-radius: 3px;
    letter-spacing: .03em;
    box-shadow: 0 2px 4px rgba(0, 0, 0, 0.1);
    font-weight: 900;
    color: #4f5254;
    width: 90%;

    &:focus {
      border-color: #488bf8;
    }
  }

  .suggestion-container {
    z-index: 999;
    margin-top: 5px;
    background-color: #fefefe;
    position: absolute;
    width: 100%;
    -webkit-box-shadow: 0 0 15px 1px rgba(69, 65, 78, 0.2);
    box-shadow: 0 0 15px 1px rgba(69, 65, 78, 0.2);
    border-radius: 4px;
    padding: 5px 0;
    color: #575962;

    .suggestion {
      cursor: pointer;
      padding: 10px 15px;
      font-size: 14px;
      font-weight: 500;

      &:hover {
        background-color: $background-color;
      }
    }
  }
}
