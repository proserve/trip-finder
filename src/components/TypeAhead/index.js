import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import './type-ahead.scss';

export default class TypeAhead extends React.Component {
  constructor(props) {
    super(props);
    const { suggestions } = this.props;
    this.state = {
      tempVal: '',
      suggestions,
      showSuggestions: false,
    };
    this.onValueChange = this.onValueChange.bind(this);
    this.handleClickOutside = this.handleClickOutside.bind(this);
  }

  componentDidMount() {
    document.addEventListener('mousedown', this.handleClickOutside);
  }

  componentWillReceiveProps({ suggestions }) {
    this.setState({ suggestions });
  }

  componentWillUnmount() {
    document.removeEventListener('mousedown', this.handleClickOutside);
  }

  onValueChange(e) {
    const { value } = e.target;
    const { suggestions } = this.props;
    const newSuggestions = suggestions.filter(
      suggestion => suggestion.toLocaleLowerCase()
        .indexOf(value.toLowerCase()) !== -1,
    );
    this.setState({ suggestions: newSuggestions, tempVal: value, showSuggestions: true });
  }

  handleClickOutside(event) {
    if (this.wrapperRef && !this.wrapperRef.contains(event.target)) {
      this.setState({ showSuggestions: false });
    }
  }


  render() {
    const { placeholder, onValueChange, style } = this.props;
    const { tempVal, suggestions, showSuggestions } = this.state;
    return (
      <div
        className="type-ahead-container"
        style={style}
        ref={(ref) => { this.wrapperRef = ref; }}
      >
        <input
          onFocus={() => this.setState({ showSuggestions: true })}
          type="text"
          value={tempVal}
          placeholder={placeholder}
          onChange={this.onValueChange}
        />
        {showSuggestions > 0 && (
          <div className="suggestion-container">
            {(suggestions && suggestions.length === 0) ? (
              <div className="suggestion">No results</div>
            ) : (
              <Fragment>
                {suggestions.map(suggestion => (
                  <div
                    onClick={() => {
                      onValueChange(suggestion);
                      this.setState({ tempVal: suggestion, showSuggestions: false });
                    }}
                    role="presentation"
                    key={suggestion}
                    className="suggestion"
                  >
                    {suggestion}
                  </div>
                ))}
              </Fragment>
            )
            }
          </div>
        )}
      </div>

    );
  }
}

TypeAhead.propTypes = {
  onValueChange: PropTypes.func.isRequired,
  style: PropTypes.objectOf(PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ])),
  placeholder: PropTypes.string,
  suggestions: PropTypes.arrayOf(PropTypes.string).isRequired,
};

TypeAhead.defaultProps = {
  placeholder: 'Type to see suggestions',
  style: {},
};
