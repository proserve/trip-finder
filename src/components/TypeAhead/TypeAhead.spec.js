import React from 'react';
import { shallow } from 'enzyme';
import TypeAhead from './index';
import data from '../../assets/response';
import TFGraph from '../../common/TFGraph';

const cities = Object.keys(new TFGraph(data.deals).graph);

describe('TypeAhead', () => {
  it('should render correctly in "debug" mode', () => {
    const component = shallow(<TypeAhead suggestions={cities} onValueChange={() => {}} />);
    expect(component).toMatchSnapshot();
  });
});
