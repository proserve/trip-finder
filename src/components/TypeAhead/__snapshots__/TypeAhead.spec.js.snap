// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TypeAhead should render correctly in "debug" mode 1`] = `
<div
  className="type-ahead-container"
  style={Object {}}
>
  <input
    onChange={[Function]}
    onFocus={[Function]}
    placeholder="Type to see suggestions"
    type="text"
    value=""
  />
</div>
`;
